qualtran/bloqs/gf_arithmetic/gf2_inverse.py

```diff
@@ -0,0 +1,167 @@
+    from qualtran.simulation.classical_sim import ClassicalValT
+
+
+@attrs.frozen
+class GF2Inverse(Bloq):
+    r"""Out of place inversion for elements in GF($2^m$)
+
+    Given a quantum register storing elements from GF($2^m$), this bloq computes the inverse
+    of the given element in a new output register, out-of-place. Specifically,
+    it implements the transformation
+
+    $$
+        |a\rangle |0\rangle \rightarrow |a\rangle |a^{-1}\rangle
+    $$
+
+    Inverse is computed by using Fermat's little theorem for Finite Fields, which states that
+    for a finite field $\mathbb{F}$ with $m$ elements, $\forall a \in \mathbb{F}$
+    $$
+        a^{m} = a
+    $$
+
+    When the finite field is GF($2^m$), Fermat's little theorem can be used to obtain the relation
+
+    $$
+        a^{-1} = a^{2^m - 2}
+    $$
+
+    Thus, the inverse can be obtained via $m - 1$ squaring and multiplication operations.
+
+    Args:
+        bitsize: The degree $m$ of the galois field $GF(2^m)$. Also corresponds to the number of
+            qubits in the input register whose inverse should be calculated.
+
+    Registers:
+        x: Input THRU register of size $m$ that stores elements from $GF(2^m)$.
+        result: Output RIGHT register of size $m$ that stores $x^{-1}$ from $GF(2^m)$.
+        junk: Output RIGHT register of size $m$ and shape ($m - 2$) that stores
+            results from intermediate multiplications.
+    """
+
+    bitsize: SymbolicInt
+
+    @cached_property
+    def signature(self) -> 'Signature':
+        junk_reg = (
+            [Register('junk', dtype=self.qgf, shape=(self.bitsize - 2,), side=Side.RIGHT)]
+            if is_symbolic(self.bitsize) or self.bitsize > 2
+            else []
+        )
+        return Signature(
+            [
+                Register('x', dtype=self.qgf),
+                Register('result', dtype=self.qgf, side=Side.RIGHT),
+                *junk_reg,
+            ]
+        )
+
+    @cached_property
+    def qgf(self) -> QGF:
+        return QGF(characteristic=2, degree=self.bitsize)
+
+    def my_static_costs(self, cost_key: 'CostKey'):
+        from qualtran.resource_counting import QubitCount
+
+        if isinstance(cost_key, QubitCount):
+            return self.signature.n_qubits()
+
+        return NotImplemented
+
+    def build_composite_bloq(self, bb: 'BloqBuilder', *, x: 'Soquet') -> Dict[str, 'Soquet']:
+        if is_symbolic(self.bitsize):
+            raise DecomposeTypeError(f"Cannot decompose symbolic {self}")
+        result = bb.allocate(dtype=self.qgf)
+        if self.bitsize == 1:
+            x, result = bb.add(GF2Addition(self.bitsize), x=x, y=result)
+            return {'x': x, 'result': result}
+
+        x = bb.add(GF2Square(self.bitsize), x=x)
+        x, result = bb.add(GF2Addition(self.bitsize), x=x, y=result)
+
+        junk = []
+        for i in range(2, self.bitsize):
```

Comment:
nit: I would argue that `range(self.bitsize - 2)` is clearer but it's up to you

Reply:
I prefer this because $i$ has a significance (we are computing $x^{2^{i}}$ in the $i$'th iteration), so I'll keep it as is.